src/irmin/tree.ml

```diff
@@ -1258,76 +1434,199 @@ module Make (P : Private.S) = struct
+            "@[<v 2>Tree.export: added inconsistent node binding@,\
+             key: %a@,\
+             value: %a@,\
+             computed hash: %a@]" pp_node_key key Node.pp_value v pp_hash h'
+      in
+      Log.err (fun f ->
+          f "Added binding %a -> %a" pp_node_key key dump (`Node n));
+      Node.export ?clear repo n key;
+      k ()
+    in
+
+    let add_node_map n (x : Node.map) k =
+      let node =
+        (* Since we traverse in post-order, all children of [x] have already
+           been added. Thus, their keys are cached and we can retrieve them. *)
+        StepMap.to_seq x
+        |> Seq.map (fun (step, v) ->
+               match v with
+               | `Node n -> (
+                   match Node.cached_key n with
+                   | Some k -> (step, `Node k)
+                   | None ->
+                       assertion_failure
+                         "Encountered child node value with uncached key \
+                          during export:@,\
+                          @ @[%a@]" dump v)
+               | `Contents (c, m) -> (
+                   match Contents.cached_key c with
+                   | Some k -> (step, `Contents (k, m))
+                   | None ->
+                       assertion_failure
+                         "Encountered child contents value with uncached key \
+                          during export:@,\
+                          @ @[%a@]" dump v))
+        |> P.Node.Val.of_seq
+      in
+      add_node n node k
+    in
+
+    let add_updated_node n (v : Node.value) (updates : Node.updatemap) k =
+      let node =
+        StepMap.fold
+          (fun k v acc ->
+            match v with
+            | Node.Remove -> P.Node.Val.remove acc k
+            | Node.Add (`Node n as v) -> (
+                match Node.cached_key n with
+                | Some ptr -> P.Node.Val.add acc k (`Node ptr)
+                | None ->
+                    assertion_failure
+                      "Encountered child node value with uncached key during \
+                       export:@,\
+                       @ @[%a@]" dump v)
+            | Add (`Contents (c, m) as v) -> (
+                match Contents.cached_key c with
+                | Some ptr -> P.Node.Val.add acc k (`Contents (ptr, m))
+                | None ->
+                    assertion_failure
+                      "Encountered child contents value with uncached key \
+                       during export:@,\
+                       @ @[%a@]" dump v))
+          updates v
+      in
+      add_node n node k
     in
-    let rec on_node (`Node n) k =
+
+    let rec on_node : type r. [ `Node of node ] -> (unit, r) cont_lwt =
+     fun (`Node n) k ->
+      (* Invariant: when [k] is called, [Node.cached_key n = Some _]. *)
+      let k () =
+        (match Node.cached_key n with
+        | None ->
+            Log.err (fun f ->
+                f "After trying to export %a, key isn't cached." dump (`Node n));
+            assert false
```

Comment:
maybe you can use the `assertion_failure` exn here too

Reply:
Good idea